core/src/main/scala/kafka/coordinator/group/CoordinatorLoaderImpl.scala

```diff
@@ -0,0 +1,168 @@
+  deserializer: Deserializer[T],
+  loadBufferSize: Int
+) extends CoordinatorLoader[T] with Logging {
+  private val isRunning = new AtomicBoolean(true)
+  private val scheduler = new KafkaScheduler(1)
+  scheduler.startup()
+
+  /**
+   * Loads the coordinator by reading all the records from the TopicPartition
+   * and applying them to the Replayable object.
+   *
+   * @param tp          The TopicPartition to read from.
+   * @param coordinator The object to apply records to.
+   */
+  override def load(
+    tp: TopicPartition,
+    coordinator: CoordinatorPlayback[T]
+): CompletableFuture[Void] = {
+    val future = new CompletableFuture[Void]()
+    val result = scheduler.scheduleOnce(s"Load coordinator from $tp",
+      () => doLoad(tp, coordinator, future))
+    if (result.isCancelled) {
+      future.completeExceptionally(new RuntimeException("Coordinator loader is closed."))
+    }
+    future
+  }
+
+  private def doLoad(
+    tp: TopicPartition,
+    coordinator: CoordinatorPlayback[T],
+    future: CompletableFuture[Void]
+  ): Unit = {
+    try {
+      replicaManager.getLog(tp) match {
+        case None =>
+          future.completeExceptionally(new NotLeaderOrFollowerException(
+            s"Could not load records from $tp because the log does not exist."))
+
+        case Some(log) =>
+          def logEndOffset: Long = replicaManager.getLogEndOffset(tp).getOrElse(-1L)
+
+          // buffer may not be needed if records are read from memory
+          var buffer = ByteBuffer.allocate(0)
+          // loop breaks if leader changes at any time during the load, since logEndOffset is -1
+          var currOffset = log.logStartOffset
+          // loop breaks if no records have been read, since the end of the log has been reached
+          var readAtLeastOneRecord = true
+
+          while (currOffset < logEndOffset && readAtLeastOneRecord && isRunning.get) {
+            val fetchDataInfo = log.read(
+              startOffset = currOffset,
+              maxLength = loadBufferSize,
+              isolation = FetchIsolation.LOG_END,
+              minOneMessage = true
+            )
+
+            readAtLeastOneRecord = fetchDataInfo.records.sizeInBytes > 0
+
+            val memoryRecords = (fetchDataInfo.records: @unchecked) match {
+              case records: MemoryRecords =>
+                records
+
+              case fileRecords: FileRecords =>
+                val sizeInBytes = fileRecords.sizeInBytes
+                val bytesNeeded = Math.max(loadBufferSize, sizeInBytes)
+
+                // minOneMessage = true in the above log.read means that the buffer may need to
+                // be grown to ensure progress can be made.
+                if (buffer.capacity < bytesNeeded) {
+                  if (loadBufferSize < bytesNeeded)
+                    warn(s"Loaded metadata from $tp with buffer larger ($bytesNeeded bytes) than " +
+                      s"configured buffer size ($loadBufferSize bytes).")
```

Comment:
Is there a reason we removed the config name here? Just for readability?

Reply:
This component is generic so I thought that having a group coordinator specific config in the log is not great.